### DISCUSSION TITLE
Basic real time plotting

### DISCUSSION THREAD

**Discussion by saturn77** [slint-ui/slint#381] [Q&A]
I am a long time real time controls engineer, often using FPGA + SoC in conjunction with telemetry link to a Qt GUI where I have typically used PyQtGraph for real time plotting and in some cases C++. 

I like the sixtyfps project, particularly as I have been writing applications in Rust for the past three years and really like its many advantages. I really like the native integration of Rust into SixtyFPS. 

So I have a new project, in which I need to do real time plotting. Any suggestions on the applicability of doing that with SixtyFPS? I would need to plot say 2 variables at any one time, and it would be nice to expand or contract the "scene" like one can do in PyQtGraph. If anyone is working on something like this, I would be happy to help out or test. Thanks. James

**Comment by tronical** [comment]
Hi @saturn77,

Generally this is a use-case that we'd like to support - it has come up a few times in the past. There's no out-of-the-box solution right now, but I'm a curious about your use-case to see what the best path forward is. So I have a few questions :-)

Is the amount of data and the type of visualisation well suited for generating the graphical representation on the CPU, or is this something that envision requiring a GPU?

If the former is acceptable, then perhaps there's a rust crate out there that can generate suitable graphics (maybe https://lib.rs/keywords/plotting has something?) and all we need to provide is a way to allow doing the controls and feeding the generated image into SixtyFPS, as an image. We have `sixtyfps::Image::load_from_path` and could have a `sixtyfps::Image::load_from_argb` function next to it perhaps.

If you require rendering on the GPU, then it depends whether this can be done with existing path rendering primitives or whether it requires custom shaders.

If you can transform your data into meaningful path curves then we could perhaps either offer an imperative "Canvas" style API ("here's a canvas to paint on, it's an FBO") or the existing Path primitives with a bit more work. The iot-dashboard example tries the latter approach.

If you're best off somehow feeding your data into a texture (or some other kind of buffer) and do the fragment generation entirely using custom shaders, then we need to provide hooks for that - even if they're initially maybe OpenGL specific.

So I'm curious, are there already good supporting plotting crates in the Rust ecosystem that we could connect with somehow?

**Reply by saturn77** [reply]
Thanks for the feedback. Overall CPU rendering would work in my applications, and using something like plotters library at https://github.com/38/plotters would suite my application space. 

From a higher level perspective, there is telemetry at say 100 Hz or 10 Hz that is common, such as in motor control or power electronics applications, often sent via USB serial. For this general class of applications, plotters would work just fine where it renders an image, and then sixtyfps renders the image using something like the sixtyfps::Image::load_from_path . . . where there is additional API work that I know would need to be done in sixtyfps to make this all work. 

So a question would be how to setup something like using serial.rs or tokio.rs -> acquire data, plotter -> create plot, and sixtyfps -> plot the data with controls. Thanks. James

**Reply by tronical** [reply]
Yes, that summarises it well. Plotters looks like a very comprehensive crate.

Regarding the setup of data flow and how to feed data from async functions into SixtyFPS: Due to windowing system restrictions the UI has to live in the main thread. What usually works then is to create the UI there, where `MyUI::new()` returns you an object that implements the `sixtyfps::ComponentHandle` trait. That one acts like a strong reference count, and you can get a weak one using `as_weak()`, which is safe to pass through threads. Once you receive data in your callback from tokio, for example, and you kept clones of the weak UI handle around, then you can call [`upgrade_in_event_loop()`](https://sixtyfps.io/snapshots/master/docs/rust/sixtyfps/struct.weak#method.upgrade_in_event_loop) (or [`sixtyfps::invoke_from_event_loop`](https://sixtyfps.io/snapshots/master/docs/rust/sixtyfps/fn.invoke_from_event_loop.html) if you stick to the latest release) to invoke a closure in the UI thread where you can send in the final data (image data or path to generated .png on disk).

You can also have a look at [cargo-ui](https://github.com/sixtyfpsui/cargo-ui/blob/master/src/main.rs) for inspiration, as that's using tokio.rs to feed data into SixtyFPS.

Edit: Filed #387 to track `sixtyfps::Image::load_from_data` (name not set in stone, of course).

**Reply by tronical** [reply]
To build on that discussion, I played a bit with plotters and feeding data into SixtyFPS and I must say: plotters is great :)

I've taken a first stab at the SixtyFPS side in https://github.com/sixtyfpsui/sixtyfps/pull/392 , which includes a little example that uses plotter to draw a scene.

**Reply by DaMilyutin** [reply]
@tronical , I want to do similar thing from C++ side. I mean fill Image's SharedPixelBuffer and send it to Image.
However, there is no big choice of constructors in Image C++ API.
Can you clarify how can I do it?

P.S. What is least painful way to switch from sixtyfps to slint?

**Reply by tronical** [reply]
The corresponding constructors (or factory functions) need to be added to slint::Image in C++. I think it would be similar to the Rust from_foo functions but maybe just take a span of bytes. There might not be a perfect data type, so it’ll likely be more low level.

P.S.: I made two migrantions, one in Rust and one in C++ and it was mostly search and replace. The biggest changes apart from that were in the Rust model API and the CMake API. Let us know if there is anything missing in the migration guides :)

**Reply by DaMilyutin** [reply]
@tronical, should I create issue for this augmented interface of Image?

Can you clarify how ownership goes bettween Rust/C++ boundary?
Can I allocate something in C++ and pass it to Rust like (Ex. `std::vector<char>` or raw pointer to allocated array of `char`, or `unique_ptr<char[]>`)?
Or must I first create Rust data owner (allocated on Rust's side) wrapped by C++ handle object (ex. SharedBuffer) and access it on C++ side?

**Reply by ogoffart** [reply]
There is already an issue for it: https://github.com/slint-ui/slint/issues/616

**Reply by ogoffart** [reply]
> Can you clarify how ownership goes bettween Rust/C++ boundary?

Everything that is allocated by the rust allocator must be freed by the rust allocator.  And similarily for the C++ allocator.
We have [SharedVector](https://docs.rs/slint/latest/slint/struct.SharedVector.html) that can be shared between rust and C++ because it only use the rust allocator.  
The SharedPixelBuffer Rust type contains a SharedVector<Pixel>,  so that's what we need in our binary interface.

**Comment by saturn77** [comment]
That is very good news. Plotters does indeed seem to be a good library, and having it connected to sixtyfps would allow for a host of applications . . .

**Comment by saturn77** [comment]
Just a quick question/comment here on the merge request  #392  for the code that you have with plotters.rs - it look like wasm bindgen attribute is being used, so I take it this example will run in the browser, that is very cool. I think I saw somewhere that people were saying sixty is for desktop and embedded, but I don't agree with that with respect to wasm tools. After all, that is how Yew and egui also do their webviews . . .

**Reply by tronical** [reply]
In principle yes, this should work in the browser, too. In practice plotters panics when trying to measure text for axis text as font-kit won’t report any fonts (or something like that). Nothing insurmountable though.

**Reply by saturn77** [reply]
I see that #392 has been merged into master now, thank you for all the hard work and various commits to make it happen. I will use the example and port to my new design.

**Comment by DaMilyutin** [comment]
I'm interested in plotting, too. Not real time though.
But I'd probably plot big arrays of data and I need to analyse those in depth.
So I need pan/zoom and to render it fast.
I'm curious about having something similar in functionality to QCustomPlot.
I wonder if this kind of widget is implementable via plotters?

**Comment by TorBorve** [comment]
I  have a very similar need as @saturn77. Given that the discussion was stated a few years ago I am curios to hear what the current state of real time plotting is? 

I have several signals which are sampled at 1kHz. I want a GUI where I can choose a plot layout (how many columns and rows). Then I want to be able to add and remove signals from the different plots. Scaling of axis would also be very nice. Some examples of similar tools I have used is [rqt_plot](http://wiki.ros.org/rqt_plot) for ROS1 and [Data Inspector](https://se.mathworks.com/help/simulink/slref/simulationdatainspector.html) for MATLAB. Would you recommend using slint with plotters for this? If so, how? Do you have any other recommendations for achieving this?

Thanks!

**Reply by BenMoon** [reply]
Hey,
I just came across this: https://github.com/slint-ui/slint/tree/master/examples/plotter
Does that help?

**Reply by TorBorve** [reply]
Hi,
Yes, thank you! I ended up doing it simple and using [PlotJuggler](https://github.com/facontidavide/PlotJuggler). Though it would be interesting to do it using Slint some time.

**Comment by okhsunrog** [comment]
@tronical Hi, is it supposed to work only in browser, or on desktop / embedded platforms too? https://github.com/slint-ui/slint/tree/master/examples/plotter

**Reply by ogoffart** [reply]
This also work on desktop and embedded platform too (that support std)

**Comment by okhsunrog** [comment]
@saturn77 
I made this demo with realtime plotting in Slint using WGPU, you might want to take a look: https://github.com/okhsunrog/slint_realtime_plotting_experiments
